src/webgpu/api/operation/render_pipeline/primitive_topology.spec.ts

```diff
@@ -0,0 +1,421 @@
+    });
+  }
+
+  makeVertexBuffer(vertexArray: Float32Array): GPUBuffer {
+    const vertexBuffer = this.device.createBuffer({
+      mappedAtCreation: true,
+      size: vertexArray.byteLength,
+      usage: GPUBufferUsage.VERTEX,
+    });
+    const vertexMapping = vertexBuffer.getMappedRange();
+    new Float32Array(vertexMapping).set(vertexArray);
+    vertexBuffer.unmap();
+    return vertexBuffer;
+  }
+
+  static generateVertexBuffer(vertexLocations: Point2D[]): Float32Array {
+    const vertexCoords = new Float32Array(vertexLocations.length * 4);
+    for (let i = 0; i < vertexLocations.length; i++) {
+      const point = vertexLocations[i].toNDC();
+      vertexCoords[i * 4 + 0] = point.x;
+      vertexCoords[i * 4 + 1] = point.y;
+      vertexCoords[i * 4 + 2] = point.z;
+      vertexCoords[i * 4 + 3] = point.w;
+    }
+    return vertexCoords;
+  }
+
+  makeIndexBuffer(indexArray: Uint32Array): GPUBuffer {
+    const indexBuffer = this.device.createBuffer({
+      mappedAtCreation: true,
+      size: indexArray.byteLength,
+      usage: GPUBufferUsage.INDEX,
+    });
+    const indexMapping = indexBuffer.getMappedRange();
+    new Uint32Array(indexMapping).set(indexArray);
+    indexBuffer.unmap();
+    return indexBuffer;
+  }
+
+  run(
+    primitiveTopology: GPUPrimitiveTopology,
+    testLocations: TestLocation[],
+    usePrimitiveRestart: boolean
+  ): void {
+    const colorAttachment = this.makeAttachmentTexture();
+
+    // Color load operator will clear color attachment to zero.
+    const encoder = this.device.createCommandEncoder();
+    const renderPass = encoder.beginRenderPass({
+      colorAttachments: [
+        {
+          attachment: colorAttachment.createView(),
+          loadValue: { r: 0.0, g: 0.0, b: 0.0, a: 0.0 },
+        },
+      ],
+    });
+
+    // Draw a primitive using 6 vertices based on the type.
+    // Pixels are generated based on vertex position.
+    // If point, 1 pixel is generated at each vertex location.
+    // Otherwise, >1 pixels could be generated.
+    const vertexModule = this.makeShaderModule('vertex', {
+      glsl: `#version 450
+            layout(location = 0) in vec4 pos;
+            void main() {
+                gl_Position = pos;
+                gl_PointSize = 1.0;
+            }`,
+    });
+
+    // Output color is solid green.
+    const fragmentModule = this.makeShaderModule('fragment', {
+      glsl: `#version 450
+            layout(location = 0) out vec4 fragColor;
+            void main() {
+              fragColor = vec4(0.0, 1.0, 0.0, 1.0);
+            }`,
+    });
+
+    renderPass.setPipeline(
+      this.device.createRenderPipeline({
+        vertexStage: { module: vertexModule, entryPoint: 'main' },
+        fragmentStage: { module: fragmentModule, entryPoint: 'main' },
+        primitiveTopology,
+        rasterizationState: {
+          frontFace: 'ccw',
+          cullMode: 'none',
+        },
```

Comment:
These are defaults; rasterizationState can be omitted entirely
```suggestion
```

Reply:
Done